packages/react-dom/src/server/ReactDOMServerFormatConfig.js

```diff
@@ -289,8 +290,65 @@ export function pushTextInstance(
     // Empty text doesn't have a DOM node representation and the hydration is aware of this.
     return;
   }
-  // TODO: Avoid adding a text separator in common cases.
-  target.push(stringToChunk(encodeHTMLTextNode(text)), textSeparator);
+  // We use a null charater to wrap every text chunk. we use this to boundary check whether we need
+  // to insert a Node textSeparator like <!-- --> when flushing segments between sets of chunks.
+  if (lastPushedText) {
+    target.push(textSeparator);
+  }
+  lastPushedText = true;
+  target.push(stringToChunk(encodeHTMLTextNode(text)));
+}
+
+// The Text Embedding flags allow the format config to know when to include or exclude
+// text separators between parent and child segments. We track the leading edge and
+// trailing edges separately however due to the fact that peeking ahead is usually
+// not possible we discern between knowing we need a separator at the leading edge and
+// assuming we might need one at the trailing edge. In the future we could get more
+// advanced with the tracking and drop the trailing edge in some cases when we know a segment
+// is followed by a Text or non-Text Node explicitly vs followed by another Segment
+export opaque type TextEmbedding = number;
+type Embeddable = {textEmbedding: TextEmbedding, ...};
+
+const NO_TEXT_EMBED = /*                 */ 0b0000;
+
+const LEADING_SEPARATOR_NEEDED = /*      */ 0b0011;
+const LEADING_TEXT_EMBED_KNOWN = /*      */ 0b0001;
+// const LEADING_TEXT_EMBED_POSSIBLE = /*   */ 0b0010;
+
+const TRAILING_SEPARATOR_NEEDED = /*     */ 0b1100;
+// const TRAILING_TEXT_EMBED_KNOWN = /*     */ 0b0100;
+const TRAILING_TEXT_EMBED_POSSIBLE = /*  */ 0b1000;
+
+// Suspense boundaries always emit a comment node at the leading and trailing edge and thus need
+// no additional separators. we also use this for the root segment even though it isn't a Boundary
+// because it cannot have pre or post text and thus matches the same embedding semantics or Boundaries
+export function textEmbeddingForBoundarySegment(): TextEmbedding {
+  lastPushedText = false;
+  return NO_TEXT_EMBED;
+}
+
+export function textEmbeddingForSegment(): TextEmbedding {
```

Comment:
While this whole feature is DOM specific and ideally we wouldn't punish other environments - it's also really DOM that's our main (and currently only target). I wouldn't be too hurt if this feature was just built into ReactFizzServer instead. Including the lastPushedText part.

Since otherwise we leak a lot of concepts and have to make up flags etc.

Reply:
Yeah I had this thought as I kept adding more functions to the format config. glad you see it the same way and makes sense. I'll move things over so we can clean up the interface

Reply:
Are you able to just use the segment.textEmbedding flag as the running tally for lastPushedText?

Reply:
not totally, for boundaries, we need to track lastPushedText internally but we never need to push a separator at the end b/c we always have the suspense comment node. I suppose if I could know I was in a segment that was a boundary or the root I could use that as an alternative way to discriminating those cases